libs/akita/src/lib/entityStore.ts

```diff
@@ -204,28 +218,36 @@ export class EntityStore<S extends EntityState = any, EntityType = getEntityType
    *
    * @example
    *
-   * store.upsert(1, { active: true })
-   * store.upsert([2, 3], { active: true })
-   * store.upsert([2, 3], entity => ({ isOpen: !entity.isOpen}))
+   * store.upsert(1, { active: true }, (id, newState) => ({ id, ...newState, enabled: true }))
+   * store.upsert([2, 3], { active: true }, (id, newState) => ({ id, ...newState, enabled: true }))
+   * store.upsert([2, 3], entity => ({ isOpen: !(entity?.isOpen ?? true) }), (id, newState) => ({ id, ...newState, enabled: true }))
    *
    */
   @transaction()
-  upsert(ids: OrArray<IDType>, newState: Partial<EntityType> | EntityType | UpdateStateCallback<EntityType> | EntityType[], options: { baseClass?: Constructor } = {}) {
+  upsert<NewEntityType extends Partial<EntityType>>(
+    ids: OrArray<IDType>,
+    newState: UpsertStateCallback<EntityType, NewEntityType> | NewEntityType,
+    onCreate: CreateStateCallback<EntityType, NewEntityType, IDType>,
+    options: { baseClass?: Constructor } = {}
+  ) {
     const toArray = coerceArray(ids);
-    const predicate = isUpdate => id => hasEntity(this.entities, id) === isUpdate;
-    const isClassBased = isFunction(options.baseClass);
+    const predicate = (isUpdate) => (id) => hasEntity(this.entities, id) === isUpdate;
+    const baseClass = options.baseClass;
+    const isClassBased = isFunction(baseClass);
     const updateIds = toArray.filter(predicate(true));
-    const newEntities = toArray.filter(predicate(false)).map(id => {
-      let entity = isFunction(newState) ? newState({} as EntityType) : newState;
+    const newEntities = toArray.filter(predicate(false)).map((id) => {
+      const oldStateObj = this.entities[id as any];
+      const newStateObj = typeof newState === 'function' ? newState(oldStateObj) : newState;
```

Comment:
- You pass the `newStateObj` to `onCreate` but it's not always the complete state as you can pass a partial update.
- I can't see the point of passing the `newStateObj` to `onCreate` as it will always be `undefined`.

Reply:
The `newState` callback returns a partial update for the old state. If no old state exists the new state is passed in the `onCreate` callback to initialize the new entity, otherwise it is directly applied to the store.

But you indicated a mistake on my part: I forgot to check if 'onCreate' needs to be called.

Reply:
Ok, I think there is another error. I will rework the code.

Reply:
> * I can't see the point of passing the `newStateObj` to `onCreate` as it will always be `undefined`. ( in a case of a function )

I have ignored that all existing entities were filtered out before. Now it should be fine.